bson/src/main/org/bson/codecs/pojo/PojoBuilderHelper.java

```diff
@@ -43,51 +47,100 @@
-                fieldTypeParameterMap.put(field.getName(), typeParameterMap);
+            }
 
-                FieldModelBuilder<?> fieldModelBuilder = getFieldBuilder(field, field.getType());
-                if (parentClassTypeData != null) {
-                    specializeFieldModelBuilder(fieldModelBuilder, typeParameterMap, parentClassTypeData.getTypeParameters());
+            for (Field field : currentClass.getDeclaredFields()) {
+                propertyNames.add(field.getName());
+                PropertyMetadata<?> propertyMetadata = getOrCreateProperty(field.getName(), propertyNameMap,
+                        getTypeData(field.getGenericType(), field.getType()), propertyTypeParameterMap, parentClassTypeData,
+                        genericTypeNames, field.getGenericType());
+                if (propertyMetadata.getField() == null) {
+                    propertyMetadata.field(field);
+                    addAnnotations(propertyMetadata, field.getDeclaredAnnotations());
                 }
-
-                classModelBuilder.addField(fieldModelBuilder);
-                field.setAccessible(true);
             }
 
             parentClassTypeData = getTypeData(currentClass.getGenericSuperclass(), currentClass);
             currentClass = currentClass.getSuperclass();
         }
 
+        for (String propertyName : propertyNames) {
+            PropertyMetadata<?> propertyMetadata = propertyNameMap.get(propertyName);
+            if (propertyMetadata.isSerializable()) {
+                classModelBuilder.addProperty(createPropertyModelBuilder(propertyMetadata));
+            }
+        }
+
         reverse(annotations);
         classModelBuilder.annotations(annotations);
-        classModelBuilder.fieldNameToTypeParameterMap(fieldTypeParameterMap);
+        classModelBuilder.propertyNameToTypeParameterMap(propertyTypeParameterMap);
 
         Constructor<T> noArgsConstructor = null;
         for (Constructor<?> constructor : clazz.getDeclaredConstructors()) {
-            if (constructor.getParameterTypes().length == 0) {
+            if (constructor.getParameterTypes().length == 0 && isPublicOrProtected(constructor.getModifiers())) {
                 noArgsConstructor = (Constructor<T>) constructor;
                 noArgsConstructor.setAccessible(true);
             }
         }
-        classModelBuilder.instanceCreatorFactory(new InstanceCreatorFactoryImpl<T>(clazz.getSimpleName(), noArgsConstructor));
+
+        classModelBuilder.instanceCreatorFactory(new InstanceCreatorFactoryImpl<T>(new CreatorExecutable<T>(clazz, noArgsConstructor)));
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T, S> PropertyMetadata<T> getOrCreateProperty(final String propertyName,
+                                                                  final Map<String, PropertyMetadata<?>> propertyNameMap,
+                                                                  final TypeData<T> typeData,
```

Comment:
I'm not quite seeing that use case.  If the public API is all ObjectId there's no point in storing as a private String field.  I could see this:

    private String id; 
    public String getId() {
        return id;
    }

    public void setId(final String id) {
        this.id = id;
    }

and wanting to serialize id as an ObjectId.  But that's a different case.  Let's discuss.